sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/Mode.scala

```diff
@@ -70,10 +74,22 @@ case class Mode(
     buffer
   }
 
-  override def eval(buffer: OpenHashMap[AnyRef, Long]): Any = {
-    if (buffer.isEmpty) {
+  override def eval(buff: OpenHashMap[AnyRef, Long]): Any = {
+    if (buff.isEmpty) {
       return null
     }
+    val buffer = if (isCollatedString(child)) {
+      val modeMap = buff.toSeq.groupMapReduce {
+        case (key: String, _) =>
+          CollationFactory.getCollationKey(UTF8String.fromString(key), collationId)
+        case (key: UTF8String, _) =>
+          CollationFactory.getCollationKey(key, collationId)
+        case (key, _) => key
+      }(x => x)((x, y) => (x._1, x._2 + y._2)).values
+      modeMap
+    } else {
+      buff
+    }
```

Comment:
Going back to the original issue (why Mode doesn't work already, while otherwise Aggregation generally works with collated strings in Spark), here's what I'm interested in: why does `PhysicalDataType.ordering(child.dataType).asInstanceOf[Ordering[AnyRef]]` not work here automatically?

afaik, ordering for `PhysicalStringType` is defined correctly:
```
private[sql] val ordering = CollationFactory.fetchCollation(collationId).comparator.compare(_, _)
```

so one would naturally expect Mode to work "as is"

did you investigate this maybe?

Reply:
@uros-db Addressing the initial question about why Mode isn't functioning with collated strings in Spark as expected, even though Aggregation typically works: It seems that while the ordering for PhysicalStringType is correctly established using `private[sql] val ordering = CollationFactory.fetchCollation(collationId).comparator.compare(_, _)`, this does not automatically resolve the issue with Mode. To illustrate, consider the example of UTF8_BINARY_LCASE where an input like `Map("a" -> 3L, "b" -> 2L, "B" -> 2L)` results in evaluating the maximum over the tuples `(2L, "B"), (2L, "b"), (3L, "a")` rather than the expected `(3L, "a"), (4L, "b")`. This indicates that the current approach doesn't aggregate values as required for Mode to operate correctly. Unit tests confirm that Mode otherwise won't work for such cases.